### ISSUE TITLE
`cast_possible_truncation` could leverage simple patterns

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust-clippy#12781]
### Summary

When a cast is followed by a trivial check, either by a pattern or condition, we could avoid emitting this link since clippy can know that the cast will always succeed.

### Lint Name

cast_possible_truncation

### Reproducer

I tried this code:

```rust
#![warn(clippy::cast_possible_truncation)]
use core::num::NonZeroU8;
struct Test<const N: usize>;
impl<const N: usize> Test<N> {
    const MY_NUM: NonZeroU8 = match N {
        0 => panic!("Test<0> is not allowed"),
        1..=255 => match NonZeroU8::new(N as u8) {
            Some(n) => n,
            None => unreachable!(),
        },
        _ => panic!("Test<N> where N >= 256 is not allowed"),
    };
}
```

I saw this happen:

```
warning: casting `usize` to `u8` may truncate the value
 --> src/lib.rs:7:41
  |
7 |         1..=255 => match NonZeroU8::new(N as u8) {
  |                                         ^^^^^^^
  |
  = help: if this is intentional allow the lint with `#[allow(clippy::cast_possible_truncation)]` ...
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_possible_truncation
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(clippy::cast_possible_truncation)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: ... or use `try_from` and handle the error accordingly
  |
7 |         1..=255 => match NonZeroU8::new(u8::try_from(N)) {
  |                                         ~~~~~~~~~~~~~~~
```

Clippy should not emit this warning, since it should be able to understand from the trivial pattern 1..=255 that the cast will succeed. Note that because this occurs in a constant and `TryFrom` is unstable in constants, `TryFrom` cannot be used here.

### Version

```text
Nightly channel
Build using the Nightly version: 1.80.0-nightly
(2024-05-07 faefc618cf48bd794cbc)
```

### Additional Labels

@rustbot label +I-false-positive
